tests/saw/test_points_to_at_type.py

```diff
@@ -0,0 +1,49 @@
+from pathlib import Path
+import unittest
+from saw import *
+from saw.exceptions import VerificationError
+from saw.llvm import Contract, LLVMType, PointerType, cryptol, void
+from typing import Union
+import saw.llvm_types as ty
+
+
+class FPointsToContract(Contract):
+    def __init__(self, check_x_type : Union[PointerType, LLVMType, None]) -> None:
+        super().__init__()
+        self.check_x_type = check_x_type
+
+    def specification(self) -> None:
+        x = self.fresh_var(ty.array(2, ty.i32), "x")
+        p = self.alloc(ty.array(4, ty.i32))
+        self.points_to(p, x, check_target_type = self.check_x_type)
+
+        self.execute_func(p)
+
+        self.points_to(p, cryptol("{x} # {x}".format(x=x.name())))
+        self.returns(void)
+
+
+class PointsToAtTypeTest(unittest.TestCase):
+    def test_points_to_at_type(self):
+        connect()
+        if __name__ == "__main__": view(LogResults())
+        bcname = str(Path('tests','saw','test-files', 'points_to_at_type.bc'))
+        mod = llvm_load_module(bcname)
+
+        result = llvm_verify(mod, "f", FPointsToContract(None))
+        self.assertIs(result.is_success(), True)
+
+        # TODO RGS: For some reason, if I put this test after the next one,
```

Comment:
I honest have no idea why this is the case. For context, I cargo-culted this from the `test_points_to_at_type` test case from `saw-script`, which originally had this test case appear after the next one. But if I use this order in `galois-toolkit`:

```diff
diff --git a/tests/saw/test_points_to_at_type.py b/tests/saw/test_points_to_at_type.py
index 60512fb..7a52061 100644
--- a/tests/saw/test_points_to_at_type.py
+++ b/tests/saw/test_points_to_at_type.py
@@ -33,14 +33,12 @@ class PointsToAtTypeTest(unittest.TestCase):
         result = llvm_verify(mod, "f", FPointsToContract(None))
         self.assertIs(result.is_success(), True)
 
-        # TODO RGS: For some reason, if I put this test after the next one,
-        # the overall TestCase fails. Why?
-        result = llvm_verify(mod, "f", FPointsToContract(ty.array(2, ty.i32)))
-        self.assertIs(result.is_success(), True)
-
         with self.assertRaises(VerificationError):
             llvm_verify(mod, "f", FPointsToContract(PointerType()))
 
+        result = llvm_verify(mod, "f", FPointsToContract(ty.array(2, ty.i32)))
+        self.assertIs(result.is_success(), True)
+
         with self.assertRaises(VerificationError):
             llvm_verify(mod, "f", FPointsToContract(ty.array(3, ty.i32)))
```

Then I end up with this error:

```
$ cabal exec -- python3 -m unittest tests/saw/test_points_to_at_type.py 
E
======================================================================
ERROR: test_points_to_at_type (tests.saw.test_points_to_at_type.PointsToAtTypeTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/rscott/Documents/Hacking/Haskell/sandbox/saw-script/galois-py-toolkit/tests/saw/test_points_to_at_type.py", line 39, in test_points_to_at_type
    result = llvm_verify(mod, "f", FPointsToContract(ty.array(2, ty.i32)))
  File "/home/rscott/Documents/Hacking/Haskell/sandbox/saw-script/galois-py-toolkit/saw/__init__.py", line 396, in llvm_verify
    raise result.exception from None
  File "/home/rscott/Documents/Hacking/Haskell/sandbox/saw-script/galois-py-toolkit/saw/__init__.py", line 361, in llvm_verify
    conn.llvm_assume(module.server_name,
  File "/home/rscott/Documents/Hacking/Haskell/sandbox/saw-script/galois-py-toolkit/saw/connection.py", line 133, in llvm_assume
    LLVMAssume(self, module, function, contract, lemma_name)
  File "/home/rscott/Documents/Hacking/Haskell/sandbox/saw-script/galois-py-toolkit/saw/commands.py", line 57, in __init__
    super(LLVMAssume, self).__init__('SAW/LLVM/assume', params, connection)
  File "/home/rscott/Documents/Hacking/Haskell/sandbox/saw-script/galois-py-toolkit/virtenv/lib/python3.8/site-packages/argo_client/interaction.py", line 57, in __init__
    self.init_state = connection.protocol_state()
  File "/home/rscott/Documents/Hacking/Haskell/sandbox/saw-script/galois-py-toolkit/saw/connection.py", line 104, in protocol_state
    return self.most_recent_result.state()
  File "/home/rscott/Documents/Hacking/Haskell/sandbox/saw-script/galois-py-toolkit/virtenv/lib/python3.8/site-packages/argo_client/interaction.py", line 159, in state
    return self._result_and_state_and_out_err()[1]
  File "/home/rscott/Documents/Hacking/Haskell/sandbox/saw-script/galois-py-toolkit/virtenv/lib/python3.8/site-packages/argo_client/interaction.py", line 148, in _result_and_state_and_out_err
    raise self.process_error(exception)
saw.exceptions.VerificationError: error: types not memory-compatible:
[4 x i32]
[2 x i32]



----------------------------------------------------------------------
Ran 1 test in 0.364s

FAILED (errors=1)
```

Do you know why that might be the case?

Reply:
Argh. Even after I rearranged the order to make `python3 -m unittest tests/saw/test_points_to_at_type.py` work in isolation, it doesn't work when other test cases are ran before it. See [this CI failure](https://github.com/GaloisInc/galois-py-toolkit/pull/12/checks?check_run_id=1962338029#step:18:8) for an example:

```
E
======================================================================
ERROR: test_points_to_at_type (tests.saw.test_points_to_at_type.PointsToAtTypeTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/galois-py-toolkit/galois-py-toolkit/tests/saw/test_points_to_at_type.py", line 33, in test_points_to_at_type
    result = llvm_verify(mod, "f", FPointsToContract(None))
  File "/home/runner/work/galois-py-toolkit/galois-py-toolkit/saw/__init__.py", line 396, in llvm_verify
    raise result.exception from None
  File "/home/runner/work/galois-py-toolkit/galois-py-toolkit/saw/__init__.py", line 361, in llvm_verify
    conn.llvm_assume(module.server_name,
  File "/opt/hostedtoolcache/Python/3.9.1/x64/lib/python3.9/site-packages/argo_client/interaction.py", line 163, in result
    return self.process_result(self._result_and_state_and_out_err()[0])
  File "/opt/hostedtoolcache/Python/3.9.1/x64/lib/python3.9/site-packages/argo_client/interaction.py", line 148, in _result_and_state_and_out_err
    raise self.process_error(exception)
saw.exceptions.VerificationError: error: types not memory-compatible:
[4 x i32]
[2 x i32]



----------------------------------------------------------------------
Ran 1 test in 0.042s

FAILED (errors=1)
```

Reply:
For now, a suitable workaround is to invoke `docker restart saw-remote-api` before `python -m unittest tests/saw/test_points_to_at_type.py`. It's ugly, but it Just Works™.